### ISSUE TITLE
Create in memory cache for  _source field

### ISSUE THREAD

**Issue by sumitjainn** [elastic/elasticsearch#15767]
I am using _source field in my sort script to sort docs based on complex logic involving multi level nested fields. But accessing _source from disk for each matched document is performing poorly, giving me ~ 3 sec response times.

It would me very efficient to have a cache for this field, so that costly disk accesses can be avoided on every request. It would cost more memory, but would solve performance problem for many advanced sorting use cases.

**Comment by jpountz** [comment]
Sorry but this is exactly why we have doc values so we won't add a cache for the _source document.

**Comment by jpountz** [comment]
Maybe you could model your documents in such a way that you don't need the source anymore?

**Comment by sumitjainn** [comment]
Here's my use case: I have hotel docs which have multiple rooms and each room has day wise price and inventory.
hotel > room > priceInventory

I have filters on all three levels. Say hotels with 5 star rating, rooms with "free wifi" amenity and having inventory of 2 for dates d1-d2 and between price p1-p2. 

Now for sorting, I want to calculate minimum price from all matched nested docs. But this isn't possible with regular sort, as I cannot specify filters on multiple nested levels inside it. So I have to utilize script sorting. But inside a script accessing nested level fields is only possible with the _source field.

Doc values cannot be used here. Caching _source would improve sort efficiency drastically, and will even improve efficiency of other operations where sort needs to be loaded. Indeed, it would be very useful to have this feature for many elasticsearch users.

Mapping:
```json
        "hotel": {
			"properties": {
			   "name": {
				  "type": "string",
			   },
			   "id": {
				  "type": "string",
			   },
			   "roomTypes": {
				  "type": "nested",
				  "properties": {
					 "roomCode": {
						"type": "string",
						"index": "not_analyzed",
					 },
					 "roomType": {
						"type": "string",
						"index": "not_analyzed",
					 },
					 "amenities": {
						  "type": "string",
						  "index": "not_analyzed",
					 },
					 "invPrice": {
						  "type": "nested",
						  "properties": {
							 "date": {
								"type": "date",
								"format": "dateOptionalTime",
							 },
							 "inventory": {
								"type": "integer",
							 },
							 "price": {
								"type": "double",
							 }
						  }
					   }
				  }
			   },
			   "starRating": {
				  "type": "short",
			   }
			}
		}
```

**Comment by jpountz** [comment]
@sumitjainn You could denormalize your data in order to have a single level of nesting:

```
        "hotel": {
            "properties": {
               "name": {
                  "type": "string",
               },
               "id": {
                  "type": "string",
               },
               "roomPrices": {
                  "type": "nested",
                  "properties": {
                     "roomCode": {
                        "type": "string",
                        "index": "not_analyzed",
                     },
                     "roomType": {
                        "type": "string",
                        "index": "not_analyzed",
                     },
                     "amenities": {
                          "type": "string",
                          "index": "not_analyzed",
                     },
                     "date": {
                        "type": "date",
                        "format": "dateOptionalTime",
                     },
                     "inventory": {
                        "type": "integer",
                     },
                     "price": {
                        "type": "double",
                     }
                  }
               },
               "starRating": {
                  "type": "short",
               }
            }
        }
```

This will force to duplicate `roomType`, `amenities` , etc. for every price. But it will also make your queries much faster sinch they will have to perform a single join instead of two and you will be able to use nested sorting: https://www.elastic.co/guide/en/elasticsearch/guide/current/nested-sorting.html

**Comment by sumitjainn** [comment]
@jpountz It was one of the approaches we tried, but sadly it doesn't work. The reason is the loss of association between room and invPrice. So the query to match only hotels for rooms which have inventory on 2 dates d1 and d2 (not across different rooms), is not possible. This is direct result of the fact that the grouping facilitated though nesting of all invPrice of one room under it, has been given away, while flattening.

**Comment by jpountz** [comment]
Ok I see. I don't have a good alternative then :(

**Comment by sumitjainn** [comment]
Is it not a valid feature request then? Wouldn't it benefit to have a source cache for many use cases apart from this? Can it atleast be made open for discussion?

**Comment by jpountz** [comment]
@sumitjainn I'm open for discussion as of how to solve this kind of problem, however I am -1 on the _source cache idea. Maybe you can open a new issue that explains the problem that you are trying to solve and links to the discussion that happened here?

**Comment by sumitjainn** [comment]
@jpountz I have created the issue, but could you please elaborate on your negative vote on caching?

**Comment by jpountz** [comment]
Elasticsearch has two ways to read field values:
 - doc values to read a couple field values from many documents, which is typically useful for sorting/aggregations
 - stored fields/_source to read as many field values as you want from a couple of documents, which is typically useful to get summaries for the top matches

So it doesn't make sense to try to optimize stored fields for the use-case that doc values have been designed for. Additionally I don't see how a cache would help for this problem.

**Comment by sumitjainn** [comment]
@jpountz I have stumbled upon another issue due to non-caching of _source:

I have top level hotel doc with large number of nested docs (around 30k) with the total doc size ~ 35 MB. Now fetching any field in the response (nested or otherwise), gives me a response time of 2-6 secs with only ~20 matching docs. Response time is always > 1 sec even with <5 matching docs.

Now when I used multiple stored fields instead of _source, my response times dropped to ~ 50ms. I have tested this on machine with 16 GB free memory to be used by OS for File System cache. 

So, isn't this too much of a performance penalty to bear? Also, the es documentation is inaccurate, as it heavily recommends _source over stored fields, while not spelling out its limitations, found thus. 

As to your last point, while we have changed our schema since then,  I don't think doc values solves the use case I had mentioned. Also I don't see why caching _source wouldn't help, wouldn't reading from memory be faster than from disk?
